wagtail/admin/views/generic/mixins.py

```diff
@@ -221,24 +225,79 @@ def setup(self, request, *args, **kwargs):
         self.object = self.get_object()
         self.lock = self.get_lock()
         self.locked_for_user = self.lock and self.lock.for_user(request.user)
+        self.setup_workflow()
         super().setup(request, *args, **kwargs)
 
+    def setup_workflow(self):
+        workflow_enabled = self.model and issubclass(self.model, WorkflowMixin)
+        self.workflow_state = None
+        self.workflow_tasks = []
+        self.current_workflow_task = None
+
+        if workflow_enabled:
+            if self.object:
+                self.workflow = self.object.get_workflow()
+            else:
+                self.workflow = self.model.get_default_workflow()
+
+        self.workflow_enabled = workflow_enabled and self.workflow is not None
+
+        if not self.workflow_enabled or not self.object:
+            return
+
+        self.workflow_state = (
```

Comment:
Do we always need these (this and below), or could they be cached properties only calculated when needed?

Reply:
I think we can calculate the `workflow_tasks` later, but the rest are used pretty early in the flow to determine the available actions.